core/src/test/java/org/bitcoinj/core/SegwitAddressTest.java

```diff
@@ -90,6 +93,44 @@ public void example_p2wpkh_testnet() {
         assertEquals(bech32.toLowerCase(Locale.ROOT), address.toString());
     }
 
+    @Test
+    public void example_p2wpkh_testnet_equality_anomaly() {
+        String bech32 = "tb1qw508d6qejxtdg4y5r3zarvary0c5xw7kxpjzsx";
+
+        SegwitAddress a = SegwitAddress.fromBech32(TESTNET, bech32);
+        SegwitAddress b = SegwitAddress.fromBech32(SigNetParams.get(), bech32);
+
+        assertNotEquals(a, b);  // TODO: Passes, but probably shouldn't
+        assertEquals(a.toString(), b.toString());
+    }
+
+    @Test
+    public void example_p2wpkh_regtest() {
+        String bech32 = "bcrt1qspfueag7fvty7m8htuzare3xs898zvh30fttu2";
+
+        SegwitAddress address = SegwitAddress.fromBech32(RegTestParams.get(), bech32);
+
+        assertEquals(RegTestParams.get(), address.params);
+        assertEquals("00148053ccf51e4b164f6cf75f05d1e62681ca7132f1",
+                ByteUtils.HEX.encode(ScriptBuilder.createOutputScript(address).getProgram()));
+        assertEquals(ScriptType.P2WPKH, address.getOutputScriptType());
+        assertEquals(bech32.toLowerCase(Locale.ROOT), address.toBech32());
+        assertEquals(bech32.toLowerCase(Locale.ROOT), address.toString());
+    }
+
+    @Test
+    public void example_p2wpkh_regtest_fails() {
+        String bcrt1_bech32 = "bcrt1qspfueag7fvty7m8htuzare3xs898zvh30fttu2";
+
+        // TODO: Passes, but probably shouldn't
```

Comment:
A rethinking of how `Networks` is implemented (as part of the migration from `NetworkParameters` to `Network`) could fix this.